### PR TITLE
[ENHANCEMENT] GE-28 "Bob" columnar multibatch profiler

### PR DESCRIPTION
**Pull request by anthonyburdi**
Creating as a draft PR for preliminary discussion / review

### CODE REVIEW

File: tests/profiler/bob_user_workflow_fixture.py
Comment (alexsherstinsky) [inline tests/profiler/bob_user_workflow_fixture.py]: It would be nice to include a docstring describing the meaning of each persona ("alice", "bob", et al) in terms of their functionality.

**Review by alexsherstinsky — APPROVED** [review]
LGTM (please see a request to include docstrings describing the personals in the Python modules containing the fixtures, even though the documentation also exists in Confluence).

**Comment by alexsherstinsky** [comment]
@anthonyburdi Some tests fail, but please see PR #2779 where they might be fixed.  Thank you.

File: tests/profiler/bob_user_workflow_fixture.py
Comment (anthonyburdi) [reply tests/profiler/bob_user_workflow_fixture.py]: Great idea! Adding now.

Comment (alexsherstinsky) [reply tests/profiler/bob_user_workflow_fixture.py]: 🙇

File: tests/profiler/bob_user_workflow_fixture.py
Comment (talagluck) [inline tests/profiler/bob_user_workflow_fixture.py]:
```suggestion
      - Wants to verify that the latest batches are not so different from all of the batches last year, if they are he wants to be alerted as maybe there are data collection issues or fares have changed significantly.
```

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (talagluck) [inline tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: To which batches do we have access here? Is it only batches we pass into profiler.profile()? Is it all batches available via the data connector? Do we need to pass the data connector to the profiler?

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (talagluck) [inline tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Could we grab the value_set parameters from a subset of batches that we selected above (e.g. we selected :-6; is it possible to only get a union of value_sets for :-4, or would that need to be its own parameter)? Not that this is necessary here, just trying to get a sense of what operations are possible on a collection of parameters across batches.

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (talagluck) [inline tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Ah - this just clicked for me. We can pass anything we want to metric_domain_kwargs, and thus get metrics for domains not built by the domain builder

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (talagluck) [inline tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: I'm trying to place this in relation to the my_value_set_for_ids parameter for my_rule_for_ids. In my_value_set_for_ids, the parameter is a list of value_sets. We then use a union in the expect_configuration_builder to get a new metric/parameter/thing (the set containing all of these value_sets). In this case, the aggregation is done in the expectation_configuration_builder.

In this case, the output of the parameter is the mean of means. We do the aggregation in the parameter_builder. 

Is that right? Is there a reason why we would want to do the aggregation in the expectation_configuration_builder in one case, and in the parameter_builder in the other case? Do we specifically want to allow that flexibility?

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (talagluck) [inline tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Is the question whether we might allow some parsing language here to allow for parameterization? I think allowing the same language that we use elsewhere in the config would be fine.

In the case of taxi data (and this is maybe the way that the data connectors work for right now??), I believe that year is a string, so being able to specify that we want to convert this to an int, and then operate on it, would be great.

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (talagluck) [inline tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Does the aggregate_fn at the end need to be declared above, or could it be calculated lazily here? For instance, could someone NOT specify an aggregate_fn above, and then reference `$parameter.my_mean_parameter_for_batches_in_previous_full_year.mean` or `$parameter.my_mean_parameter_for_batches_in_previous_full_year.stdev` or `$parameter.my_mean_parameter_for_batches_in_previous_full_year.sum`? Or is the earlier caluclation necessary?

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (talagluck) [inline tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Ah. I asked about this above - thanks for clarifying here! I think we could potentially wind up with many ways of accessing the same parameters, and I would question whether we need all of these.

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (talagluck) [inline tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Just re-iterating. What batches are available here? I could imagine that rather than passing a batch into the profiler, we pass in a DataConnector. Then we review all of the BatchRequests contained in the config, and instantiate a validator containing all of them

**Review by talagluck — COMMENTED** [review]
This is awesome! Love having the artifact to review and discuss. I have a bunch of questions, and I think this could probably use some more discussion, but if you want to merge it in prior to discussion, let me know.

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (anthonyburdi) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Great question. I have been assuming that since in `Profiler.profile()` we require one of four ways to specify batch(es) to profile that these are the assumed batches, the batches a user specifies for profiling. 

One of these ways (batch_request) allows you to specify the data connector 

(docstring of Profiler.profile():
```
Utilizing one of validator, batch, batches, or batch_request (only one may be provided),
        evaluates Profiler object to evaluate rule set on the given data and returns results of rule
        evaluations as an Expectation Suite
```

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (anthonyburdi) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Hmm great question. I would assume the latter - :-4 would need to be it's own parameter given the current design.
If the parser is able to handle slices, we could do something like `union($parameter.my_value_set_for_ids.parameter_list[:-2])` to get a subset of batches when substituting the parameter (but could not get additional batches as you would have to do going from :-6 to :-4)

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (anthonyburdi) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: OH GREAT POINT! That makes total sense. I'm not sure if that example goes into Alice / Bob or another more advanced User Workflow but we should add that config soon.

Comment (anthonyburdi) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: I guess this could also work for the "I want to pin this parameter to a specific domain within the domains built by the domain builder"

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (anthonyburdi) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Ah - I was purposely trying to show both examples in this fixture (along with the third example of multiple parameters). Since both ParameterBuilders and ExpectationConfigurationBuilders are extensible I think both are valid configurations. If we want to steer toward mostly using ExpectationConfigurationBuilders for handling complex logic we can do that in the docs / examples. But I think the final call on whether this is a valid config is up to the product team.

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (anthonyburdi) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Correct - what would be an example of that syntax? `year: max(int(year))-1` to get the year previous to the latest?

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (anthonyburdi) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Hmm - great question. The way we are currently calculating the parameters is as they are defined so we would need to specify in the parameter builder. But this sounds like a great optimization step, e.g. when we talked about building the computation graph first after processing the whole config before calculating any metrics (also holding off on any metrics hidden behind conditionals until we evaluate the conditional to avoid calculating unnecessary metrics.).

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (anthonyburdi) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Interesting - so change the `Profiler.profile()` method to take a data connector instead... I think that makes sense and jibes with James' comment about unioning batches. Currently we aren't doing that yet.

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (talagluck) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: I'm not sure I understand how `max` is being used here. I wonder if we'll want a NOW() type of function though.

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (talagluck) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: That's what I'm thinking. Otherwise I can see a lot of situation where a user tries to set up batches which aren't passed in. If we're allowing BatchRequest-ish things within the config, I think this would make sense as well - otherwise the BatchRequest would only work against BatchRequests already contained with the Validator, which feels very limited.

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (anthonyburdi) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Ah I was thinking `max()` of available years from the data connector (which I'm not sure how we would access), but I also like the idea of a `now()` function. E.g. something more like `year: date_trunc(now(), 'year') - 1`

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (anthonyburdi) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: This may be silly - but would we create a new validator within the builder to handle the specific BatchRequest? I guess that would defeat the purpose of the computation tree and there is a chance we would re-compute metrics (unless we were storing them somewhere accessible to any validators)

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (alexsherstinsky) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: @anthonyburdi @talagluck I have been working under impression that there can be only one `BatchRequest` supplied to `get_validator()` at a time.  This `BatchRequest`, of course, can lead to multiple batches, including the ones whose schemas are not the same.  This is to say that so far, I have not encountered the need for `List[BatchRequest]`.  Hence, a multi-table set of batches can be retrieved with a single  `BatchRequest`.  In addition, the underlying objects, such as `DataConnector` should never be exposed in the user-facing API.

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (alexsherstinsky) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: @anthonyburdi @talagluck Why not instead use a `BatchRequest` with the query that will result in the `Batch` objects that are needed?  In other words, front-load the batch selection to `BatchRequest`, and let the `$parameter.*` syntax be used in a direct way -- we compute parameters and assign the reference to them to `$parameter.*` -- and use all of this value, without additional syntax.  Otherwise, we are getting into the realm of supporting a large syntax.  Do we have use cases to justify such increase in complexity for users to have to learn?

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (alexsherstinsky) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: @anthonyburdi @talagluck -- I think that we are saying "While domains are built only by the configured domain builder, the parameters can rely on any metric, specified by `metric_name` and `metric_domain_kwargs`." -- is this correct?

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (alexsherstinsky) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: @anthonyburdi @talagluck I believe that `parameter_builders` now require `parameter_name` (as opposed to `name`).

Also, I believe that it is correct to say that in a lot of cases, aggregation will be done in an `ExpectationConfigurationBuilder` subclass.  This is because the `build_expectation_configuration` loop is run after all the parameters have been built and are thus available.  All the batches can be processed in the parameter builders, parameters set, and then parameters queried and their values summarized/aggregated in `ExpectationConfigurationBuilder`.  It would be really interesting to try to implement a few use cases and gauge the viability and degree of convenience of this "dev loop".

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (alexsherstinsky) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Would it make sense, at least initially, to steer away from complex syntax in YML configuration files, but instead create sophisticated `ParameterBuilder` and `ExpectationConfigurationBuilder` classes to implement the specified use cases? I believe that if we get sufficient experience with this approach, then ideas for higher-level syntax will emerge.

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (alexsherstinsky) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Makes sense (though I would propose to defer this optimization till the time we can measure the actual usage patterns).

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (alexsherstinsky) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Again, as mentioned before, I would prefer to implement the `Mean` in an aggregator class first, before promoting it to the configuration syntax level.

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (alexsherstinsky) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: @talagluck @anthonyburdi I would love to discuss use cases at length.  The way I see it now, no component of `DataSource`, including `DataConnector` must be surfaced as parameters the API level.  All the smarts should be expressed via `BatchRequest` -- and it should be sufficient to express the querying for any bathes.  Should we find a limitation, then we would enrich the `BatchRequest` capabilities, not expose its processing internals.  I believe that if we describe use cases and the requirements for `Batch` objects that should be found, then we can figure out whether or not the existing `BatchRequest` expressivity is sufficient.  Thanks.

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (anthonyburdi) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Totally agree @alexsherstinsky - that is what I meant by `:-4` currently needing its own parameter (and it's own `BatchRequest` with filter). The example I gave was purely hypothetical, no use case behind it so we can leave this as a thought experiment and the code as-is for now :)

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (anthonyburdi) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: Well said @alexsherstinsky! Yes I think that is correct and we should put this note in documentation.

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (anthonyburdi) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: @alexsherstinsky 1. yes - thanks for the catch! 2. I think this is a really valid point that given the current design of calculating metrics as parameters are built a user would have to make sure to order their `ParameterBuilders` carefully if they reference other parameters. In the case where a `ParameterBuilder` does its own aggregation, the ordering is not so important.
My unsubstantiated feeling (could be wrong) is that in general we want to steer toward any aggregation in `ExpectationConfigurationBuilder`s but leave an escape hatch with custom `ParameterBuilder`s.

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (anthonyburdi) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: I think you are right @alexsherstinsky for `DomainBuilder` and `ParameterBuilder` but this is more in reference to filtering in `BatchRequest`s which I believe can already be done with custom `batch_filters` e.g. `custom_filter_function` (@talagluck would a `custom_filter_function` work for this case?)

File: tests/profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (alexsherstinsky) [reply tests/profiler/bob_user_workflow_verbose_profiler_config.yml]: @anthonyburdi FWIW, I feel the same (re aggregations should be in ECB).

File: tests/rule_based_profiler/bob_user_workflow_verbose_profiler_config.yml
Comment (alexsherstinsky) [inline tests/rule_based_profiler/bob_user_workflow_verbose_profiler_config.yml:200]: @anthonyburdi We can now shorten the module name to end at `great_expectations.profiler.parameter_builder` for the verbose, or omit it at all for the brief configuration YAML file.

**Pull request merged**
